src/main/java/games/strategy/net/ClientMessenger.java

```diff
@@ -37,39 +42,42 @@
   private volatile boolean m_shutDown = false;
 
   /**
-   * Note, the name paramater passed in here may not match the name of the
+   * Note, the name parameter passed in here may not match the name of the
    * ClientMessenger after it has been constructed.
    */
   public ClientMessenger(final String host, final int port, final String name, final String mac,
       final IConnectionLogin login) throws IOException {
     this(host, port, name, mac, new DefaultObjectStreamFactory(), login);
   }
 
+
   /**
-   * Note, the name paramater passed in here may not match the name of the
+   * Note, the name parameter passed in here may not match the name of the
    * ClientMessenger after it has been constructed.
    */
+  @VisibleForTesting
   public ClientMessenger(final String host, final int port, final String name, final String mac)
       throws IOException {
-    this(host, port, name, mac, new DefaultObjectStreamFactory());
+    this(host, port, name, mac, new DefaultObjectStreamFactory(), null);
   }
 
   /**
-   * Note, the name paramater passed in here may not match the name of the
-   * ClientMessenger after it has been constructed.
-   */
-  public ClientMessenger(final String host, final int port, final String name, final String mac,
-      final IObjectStreamFactory streamFact) throws IOException {
-    this(host, port, name, mac, streamFact, null);
-  }
-
-  /**
-   * Note, the name paramater passed in here may not match the name of the
+   * Note, the name parameter passed in here may not match the name of the
    * ClientMessenger after it has been constructed.
    */
   public ClientMessenger(final String host, final int port, final String name, final String mac,
       final IObjectStreamFactory streamFact, final IConnectionLogin login)
       throws IOException {
+    Preconditions.checkNotNull(mac);
+
+    Preconditions.checkArgument(mac.length() == 28,
+        "incorrect mac length: " + mac.length() + ", value = " + mac);
+    final String startString = MD5Crypt.MAGIC + "MH$";
+    Preconditions.checkArgument(mac.startsWith(startString),
+        "mac must start with expected start string: " + startString + ", value was: " + mac);
+    
+    Preconditions.checkArgument(ClientLoginValidator.isValidMac(mac),
+        "Not a valid mac: " + mac + ", length: " + mac.length());
```

Comment:
I feel like this code duplicates `isValidMac` in a way...

Reply:
I'm a bit confused, we invoke `isValidMac`, how are we duplicating it? While working on this code, I found that we fail *much* later due to invalid mac, if at all, and it even then is not clear that this was the problem. I added this line for error detection - "fail early" principle.

Reply:
Ah, i see what you mean now, I think when I tackled this first I added the conditions incrementally to determine the cause of failure, whether it was bad start of string or length. We are losing that info when we just fail on simply 'invalid mac'. I don't think that is a huge deal, I removed the 'extra' precondition checks as suggested.